### ISSUE TITLE
Compiling for Debian 11

### ISSUE THREAD

**Issue by GeorgelT** [irods/externals#170]
Hello, 

we're trying to setup a test instance of iRODS on our HPC cluster running debian 11. I've been having substantial issues getting to any .deb files I could use to setup a test server and test client. 

I've been following the Non-Package Install (Run In Place) guide present in the documentation [here](https://docs.irods.org/4.2.11/getting_started/installation/#non-package-install-run-in-place). I'm really having issues getting even the externals to finish compile. 

The compilation fails at avro first, then if I remove that from the Makefile, it fails at 2 other dependencies.

Here are the steps I've taken with the appropriate listings:

```
root@bjoern50:~/irods# rm -rf externals
root@bjoern50:~/irods# git clone https://github.com/irods/externals
Cloning into 'externals'...
remote: Enumerating objects: 765, done.
remote: Counting objects: 100% (498/498), done.
remote: Compressing objects: 100% (54/54), done.
remote: Total 765 (delta 466), reused 444 (delta 444), pack-reused 267
Receiving objects: 100% (765/765), 156.44 KiB | 2.14 MiB/s, done.
Resolving deltas: 100% (529/529), done.
root@bjoern50:~/irods# cd externals/
root@bjoern50:~/irods/externals# ls
ClangOverrides.txt  README.md  install_prerequisites.py                               versions.json
Makefile            build.py   irods_consortium_continuous_integration_build_hook.py
root@bjoern50:~/irods/externals# ./install_prerequisites.py 
INFO - Detected: debian
INFO - running: ['sudo', 'apt-get', 'update', '-y'], unsafe_shell=False, check_rc=getting updates failed, retries=0
INFO -   stdout: Hit:1 http://ftp.at.debian.org/debian bullseye InRelease
Get:2 https://security.debian.org/debian-security bullseye-security InRelease [44.1 kB]
Ign:3 https://www.beegfs.io/release/beegfs_7.3.0 bullseye InRelease
Get:4 https://www.beegfs.io/release/beegfs_7.3.0 bullseye Release [3378 B]
Get:5 https://www.beegfs.io/release/beegfs_7.3.0 bullseye Release.gpg [833 B]
Get:6 https://security.debian.org/debian-security bullseye-security/main Sources [121 kB]
Get:7 https://security.debian.org/debian-security bullseye-security/main i386 Packages [145 kB]
Get:8 https://security.debian.org/debian-security bullseye-security/main amd64 Packages [146 kB]
Get:9 https://security.debian.org/debian-security bullseye-security/main Translation-en [90.3 kB]
Get:10 https://www.beegfs.io/release/beegfs_7.3.0 bullseye/non-free amd64 Packages [3804 B]
Fetched 554 kB in 1s (681 kB/s)
Reading package lists...
INFO -   stderr: 
INFO - 
INFO - running: ['sudo', 'DEBIAN_FRONTEND=noninteractive', 'apt-get', 'install', '-y', 'curl', 'automake', 'make', 'autoconf2.13', 'texinfo', 'help2man', 'git', 'gpg', 'lsb-release', 'libtool', 'libbz2-dev', 'zlib1g-dev', 'libcurl4-gnutls-dev', 'libxml2-dev', 'pkg-config', 'python3-dev', 'uuid-dev', 'libssl-dev', 'fuse', 'libfuse2', 'libfuse-dev', 'libmicrohttpd-dev', 'unixodbc-dev', 'g++', 'procps'], unsafe_shell=False, check_rc=installing prerequisites failed, retries=0
INFO -   stdout: Reading package lists...
Building dependency tree...
Reading state information...
autoconf2.13 is already the newest version (2.13-69).
automake is already the newest version (1:1.16.3-2).
libbz2-dev is already the newest version (1.0.8-4).
curl is already the newest version (7.74.0-1.3+deb11u1).
libcurl4-gnutls-dev is already the newest version (7.74.0-1.3+deb11u1).
fuse is already the newest version (2.9.9-5).
libfuse-dev is already the newest version (2.9.9-5).
libfuse2 is already the newest version (2.9.9-5).
g++ is already the newest version (4:10.2.1-1).
git is already the newest version (1:2.30.2-1).
gpg is already the newest version (2.2.27-2+deb11u1).
help2man is already the newest version (1.48.1).
libmicrohttpd-dev is already the newest version (0.9.72-2).
libtool is already the newest version (2.4.6-15).
libxml2-dev is already the newest version (2.9.10+dfsg-6.7+deb11u1).
lsb-release is already the newest version (11.1.0).
make is already the newest version (4.3-4.1).
pkg-config is already the newest version (0.29.2-1).
procps is already the newest version (2:3.3.17-5).
python3-dev is already the newest version (3.9.2-3).
texinfo is already the newest version (6.7.0.dfsg.2-6).
unixodbc-dev is already the newest version (2.3.6-0.1+b1).
uuid-dev is already the newest version (2.36.1-8+deb11u1).
zlib1g-dev is already the newest version (1:1.2.11.dfsg-2+deb11u1).
The following additional packages will be installed:
  libssl1.1
The following packages will be upgraded:
  libssl-dev libssl1.1
2 upgraded, 0 newly installed, 0 to remove and 37 not upgraded.
Need to get 3372 kB of archives.
After this operation, 0 B of additional disk space will be used.
Get:1 https://security.debian.org/debian-security bullseye-security/main amd64 libssl-dev amd64 1.1.1n-0+deb11u2 [1814 kB]
Get:2 https://security.debian.org/debian-security bullseye-security/main amd64 libssl1.1 amd64 1.1.1n-0+deb11u2 [1558 kB]
apt-listchanges: Reading changelogs...
Preconfiguring packages ...
Fetched 3372 kB in 0s (10.2 MB/s)
(Reading database ... 310433 files and directories currently installed.)
Preparing to unpack .../libssl-dev_1.1.1n-0+deb11u2_amd64.deb ...
Unpacking libssl-dev:amd64 (1.1.1n-0+deb11u2) over (1.1.1n-0+deb11u1) ...
Preparing to unpack .../libssl1.1_1.1.1n-0+deb11u2_amd64.deb ...
Unpacking libssl1.1:amd64 (1.1.1n-0+deb11u2) over (1.1.1n-0+deb11u1) ...
Setting up libssl1.1:amd64 (1.1.1n-0+deb11u2) ...
locale: Cannot set LC_ALL to default locale: No such file or directory
Setting up libssl-dev:amd64 (1.1.1n-0+deb11u2) ...
Processing triggers for libc-bin (2.31-13+deb11u3) ...
INFO -   stderr: apt-listchanges: Can't set locale; make sure $LC_* and $LANG are correct!
INFO - 
INFO - running: ['sudo', 'apt-get', 'install', '-y', 'autoconf', 'rsync'], unsafe_shell=False, check_rc=installing autoconf failed, retries=0
INFO -   stdout: Reading package lists...
Building dependency tree...
Reading state information...
autoconf is already the newest version (2.69-14).
rsync is already the newest version (3.2.3-4+deb11u1).
0 upgraded, 0 newly installed, 0 to remove and 37 not upgraded.
INFO -   stderr: 
INFO - 
INFO - running: ['sudo', 'apt-get', 'install', '-y', 'patchelf'], unsafe_shell=False, check_rc=installing patchelf failed, retries=0
INFO -   stdout: Reading package lists...
Building dependency tree...
Reading state information...
patchelf is already the newest version (0.12-1).
0 upgraded, 0 newly installed, 0 to remove and 37 not upgraded.
INFO -   stderr: 
INFO - 
INFO - running: curl -sSL https://rvm.io/mpapis.asc | sudo gpg --import -, unsafe_shell=True, check_rc=curl failed, retries=0
INFO -   stdout: 
INFO -   stderr: gpg: key 3804BB82D39DC0E3: 47 signatures not checked due to missing keys
gpg: key 3804BB82D39DC0E3: "Michal Papis (RVM signing) <mpapis@gmail.com>" not changed
gpg: Total number processed: 1
gpg:              unchanged: 1
INFO - 
INFO - running: curl -sSL https://rvm.io/pkuczynski.asc | sudo gpg --import -, unsafe_shell=True, check_rc=curl failed, retries=0
INFO -   stdout: 
INFO -   stderr: gpg: key 105BD0E739499BDB: "Piotr Kuczynski <piotr.kuczynski@gmail.com>" not changed
gpg: Total number processed: 1
gpg:              unchanged: 1
INFO - 
INFO - running: curl -sSL https://get.rvm.io | sudo bash -s stable, unsafe_shell=True, check_rc=curl failed, retries=0
INFO -   stdout: Downloading https://github.com/rvm/rvm/archive/1.29.12.tar.gz
Downloading https://github.com/rvm/rvm/releases/download/1.29.12/1.29.12.tar.gz.asc
GPG verified '/usr/local/rvm/archives/rvm-1.29.12.tgz'
Upgrading the RVM installation in /usr/local/rvm/
Upgrade of RVM in /usr/local/rvm/ is complete.
<warn>Thanks for installing RVM 🙏</warn>
Please consider donating to our open collective to help us maintain RVM.

👉  Donate: <code>https://opencollective.com/rvm/donate</code>
INFO -   stderr: gpg: Signature made Fri Jan 15 19:46:22 2021 CET
gpg:                using RSA key 7D2BAF1CF37B13E2069D6956105BD0E739499BDB
gpg: Good signature from "Piotr Kuczynski <piotr.kuczynski@gmail.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 7D2B AF1C F37B 13E2 069D  6956 105B D0E7 3949 9BDB
  * Please do NOT forget to add your users to the <code>rvm</code> group.
     The installer no longer auto-adds root or users to the rvm group. Admins must do this.
     Also, please note that group memberships are ONLY evaluated at login time.
     This means that users must log out then back in before group membership takes effect!
INFO - 
INFO - running: sudo -E su -c '/usr/local/rvm/bin/rvm reload && /usr/local/rvm/bin/rvm requirements run && /usr/local/rvm/bin/rvm install 2.6.5', unsafe_shell=True, check_rc=rvm ruby install failed, retries=0
INFO -   stdout: RVM reloaded!
Checking requirements for debian.
Requirements installation successful.
INFO -   stderr: Already installed ruby-2.6.5.
To reinstall use:

    rvm reinstall ruby-2.6.5
INFO - 
INFO - running: sudo -E su -c 'PATH="/usr/local/rvm/rubies/ruby-2.6.5/bin:/usr/local/rvm /usr/local/rvm/bin/rvm/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/rvm/bin"; export PATH; /usr/local/rvm/bin/rvm reload && /usr/local/rvm/bin/rvm use 2.6.5 && gem install -v 1.14.1 fpm', unsafe_shell=True, check_rc=fpm gem install failed, retries=0
INFO -   stdout: RVM reloaded!
Successfully installed fpm-1.14.1
Parsing documentation for fpm-1.14.1
Done installing documentation for fpm after 2 seconds
1 gem installed
INFO -   stderr: Warning: PATH set to RVM ruby but GEM_HOME and/or GEM_PATH not set, see:
    https://github.com/rvm/rvm/issues/3212

Hint: To fix PATH errors try using 'rvmsudo' instead of 'sudo', see:
    https://stackoverflow.com/questions/27784961/received-warning-message-path-set-to-rvm-after-updating-ruby-version-using-rvm/28080063#28080063


Warning: PATH set to RVM ruby but GEM_HOME and/or GEM_PATH not set, see:
    https://github.com/rvm/rvm/issues/3212

Hint: To fix PATH errors try using 'rvmsudo' instead of 'sudo', see:
    https://stackoverflow.com/questions/27784961/received-warning-message-path-set-to-rvm-after-updating-ruby-version-using-rvm/28080063#28080063


Warning: PATH set to RVM ruby but GEM_HOME and/or GEM_PATH not set, see:
    https://github.com/rvm/rvm/issues/3212

Hint: To fix PATH errors try using 'rvmsudo' instead of 'sudo', see:
    https://stackoverflow.com/questions/27784961/received-warning-message-path-set-to-rvm-after-updating-ruby-version-using-rvm/28080063#28080063

RVM is not a function, selecting rubies with 'rvm use ...' will not work.

Please avoid using "sudo" in front of "rvm".
RVM knows when to use "sudo" and will use it only when it is necessary.
INFO - 
root@bjoern50:~/irods/externals# make
./build.py packagesfile
./build.py -v autoconf > autoconf.log 2>&1
./build.py -v cmake > cmake.log 2>&1
./build.py -v clang > clang.log 2>&1
./build.py -v boost > boost.log 2>&1
./build.py -v avro > avro.log 2>&1
make: *** [Makefile:26: irods-externals-avro1.11.0-1_1.0~bullseye_amd64.deb] Error 2

```
from the avro.log the excerp where the errors begin is:

```
[ 72%] ESC[32mESC[1mLinking CXX executable AvrogencppTestsESC[0m
make[3]: Leaving directory '/root/irods/externals/avro1.11.0-1_src/avro/lang/c++/build'
make[2]: Leaving directory '/root/irods/externals/avro1.11.0-1_src/avro/lang/c++/build'
make[1]: Leaving directory '/root/irods/externals/avro1.11.0-1_src/avro/lang/c++/build'
INFO -   stderr: /root/irods/externals/avro1.11.0-1_src/avro/lang/c++/impl/Stream.cc:129:27: warning: class with destructor marked 'final' cannot be inherited from [-Wfinal-dtor-non-final-class]
    ~MemoryOutputStream() final {
                          ^
/root/irods/externals/avro1.11.0-1_src/avro/lang/c++/impl/Stream.cc:120:7: note: mark 'avro::MemoryOutputStream' as 'final' to silence this warning
class MemoryOutputStream : public OutputStream {
      ^
/root/irods/externals/avro1.11.0-1_src/avro/lang/c++/impl/Stream.cc:129:27: warning: class with destructor marked 'final' cannot be inherited from [-Wfinal-dtor-non-final-class]
    ~MemoryOutputStream() final {
                          ^
/root/irods/externals/avro1.11.0-1_src/avro/lang/c++/impl/Stream.cc:120:7: note: mark 'avro::MemoryOutputStream' as 'final' to silence this warning
class MemoryOutputStream : public OutputStream {
      ^
1 warning generated.
1 warning generated.
clang-13: warning: argument unused during compilation: '-nostdinc++' [-Wunused-command-line-argument]
clang-13: warning: argument unused during compilation: '-nostdinc++' [-Wunused-command-line-argument]
clang-13: warning: argument unused during compilation: '-nostdinc++' [-Wunused-command-line-argument]
clang-13: warning: argument unused during compilation: '-nostdinc++' [-Wunused-command-line-argument]
/usr/bin/ld: libavrocpp.so.1.11.0: undefined reference to `snappy::Compress(char const*, unsigned long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*)'
/usr/bin/ld: libavrocpp.so.1.11.0: undefined reference to `snappy::Uncompress(char const*, unsigned long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*)'
clang-13: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [CMakeFiles/CompilerTests.dir/build.make:105: CompilerTests] Error 1
make[2]: *** [CMakeFiles/Makefile2:560: CMakeFiles/CompilerTests.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....
clang-13: warning: argument unused during compilation: '-nostdinc++' [-Wunused-command-line-argument]
/usr/bin/ld: libavrocpp.so.1.11.0: undefined reference to `snappy::Compress(char const*, unsigned long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*)'
/usr/bin/ld: libavrocpp.so.1.11.0: undefined reference to `snappy::Uncompress(char const*, unsigned long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*)'
clang-13: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [CMakeFiles/SpecificTests.dir/build.make:105: SpecificTests] Error 1
make[2]: *** [CMakeFiles/Makefile2:846: CMakeFiles/SpecificTests.dir/all] Error 2

```
seems to be related to snappy something...is that a different dependency?

I've tried the steps described in the documentation even on a fresh VM and there it fails sooner since it needs aditional packages that are not pulled in by the initial python script. 

I'd also really appreciate if there was a way to compile iRODS without the hardcoded externals. We have all these dependencies in the same versions as part of our cluster module system, I don't see a need to have them installed/compiled separately on the system.

I really only need some Debian 11 compatible .deb files to test this and see if we want to use it in production. At the moment it does not look like a tool we can even install.

**Comment by korydraughn** [comment]
We will look into this and report back when we have more information.

**Comment by trel** [comment]
Hi @GeorgelT, sorry for the trouble.

We've recently uploaded our builds of the externals for deb11 to packages.irods.org - in case that's helpful to get you over the externals hurdle.   We have not yet abstracted our system enough to not require building the tower of externals - required due to our need for new compilers and C++ standards and libraries on older distributions.

I've been building the current externals in a container, and have just rebuilt the avro package...

```
root@2b99ac9c43ad:/externals# cat /etc/os-release | grep PRETTY
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
root@2b99ac9c43ad:/externals# git log --oneline | head -n1
a00fced [irods/irods_rule_engine_plugin_audit_amqp#84] lib
root@2b99ac9c43ad:/externals# make avro_clean
Cleaning avro...
root@2b99ac9c43ad:/externals# make
./build.py -v avro > avro.log 2>&1
./build.py -v mungefs > mungefs.log 2>&1
root@2b99ac9c43ad:/externals# ls -l irods-externals-avro1.11.0-1_1.0~bullseye_amd64.deb 
-rw-r--r-- 1 root root 1872802 May 18 12:32 irods-externals-avro1.11.0-1_1.0~bullseye_amd64.deb
```

And upon inspection... I see this...

```
root@2b99ac9c43ad:/externals# grep snappy avro.log 
INFO -   stderr: Disabled snappy codec. libsnappy not found.
```

So, somehow, you must have libsnappy...

I tried the same...

```
root@2b99ac9c43ad:/externals# apt install -y libsnappy-dev
```
and then build again...
```
root@2b99ac9c43ad:/externals# make avro_clean
Cleaning avro...
root@2b99ac9c43ad:/externals# make avro
./build.py -v avro > avro.log 2>&1
make: *** [Makefile:26: irods-externals-avro1.11.0-1_1.0~bullseye_amd64.deb] Error 2
```

Same error as you...

```
root@2b99ac9c43ad:/externals# grep snappy avro.log 
-- Found Snappy: /usr/lib/x86_64-linux-gnu/libsnappy.so  
INFO -   stderr: Enabled snappy codec
/usr/bin/ld: libavrocpp.so.1.11.0: undefined reference to `snappy::Compress(char const*, unsigned long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*)'
/usr/bin/ld: libavrocpp.so.1.11.0: undefined reference to `snappy::Uncompress(char const*, unsigned long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*)'
```

So, the approaches appear to be...

1. uninstall `libsnappy-dev`
2. use a CMake variable to tell avro not to find snappy (not sure this is exposed/available)
3. fix avro itself, but this code hasn't changed in a while

https://github.com/apache/avro/blame/master/lang/c%2B%2B/CMakeLists.txt#L81

**Comment by GeorgelT** [comment]
Thank you for adding bullseye to targets on the .deb repository. It only had ubuntu packages until today. I hope packages will become available at some point. I'll try to remove libsnappy and see if it compiles then.

**Comment by GeorgelT** [comment]
So after uninstalling libsnappy-dev, now avro and mungefs do compile, but I'm stil left with qpid-proton failing to compile.

```
root@bjoern50:~/irods/externals# git restore Makefile
root@bjoern50:~/irods/externals# make
./build.py packagesfile
./build.py -v avro > avro.log 2>&1
./build.py -v mungefs > mungefs.log 2>&1
./build.py -v qpid-proton > qpid-proton.log 2>&1
make: *** [Makefile:178: irods-externals-qpid-proton0.36.0-0_1.0~bullseye_amd64.deb] Error 2
root@bjoern50:~/irods/externals# less qpid-proton
qpid-proton0.36.0-0_src/ qpid-proton.log          
root@bjoern50:~/irods/externals# less qpid-proton.log 

```
Excerp from the log file:
```

[ 51%] ESC[32mESC[1mLinking CXX shared library libqpid-proton-cpp.soESC[0m
make[3]: Leaving directory '/root/irods/externals/qpid-proton0.36.0-0_src/qpid-proton/build'
make[2]: Leaving directory '/root/irods/externals/qpid-proton0.36.0-0_src/qpid-proton/build'
make[1]: Leaving directory '/root/irods/externals/qpid-proton0.36.0-0_src/qpid-proton/build'
INFO -   stderr: clang-13: warning: argument unused during compilation: '-nostdinc++' [-Wunused-command-line-argument
]
/usr/bin/ld: CMakeFiles/qpid-proton-cpp.dir/src/connect_config.cpp.o: in function `proton::parse(Json::Value, proton:
:connection_options&)':
/root/irods/externals/qpid-proton0.36.0-0_src/qpid-proton/cpp/src/connect_config.cpp:165: undefined reference to `Jso
n::Value::Value(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)'
/usr/bin/ld: /root/irods/externals/qpid-proton0.36.0-0_src/qpid-proton/cpp/src/connect_config.cpp:177: undefined refe
rence to `Json::Value::asString() const'
/usr/bin/ld: /root/irods/externals/qpid-proton0.36.0-0_src/qpid-proton/cpp/src/connect_config.cpp:179: undefined refe
rence to `Json::Value::asString() const'
/usr/bin/ld: CMakeFiles/qpid-proton-cpp.dir/src/connect_config.cpp.o: in function `proton::(anonymous namespace)::par
se_sasl(Json::Value, proton::connection_options&)':
/root/irods/externals/qpid-proton0.36.0-0_src/qpid-proton/cpp/src/connect_config.cpp:112: undefined reference to `Jso
n::Value::asString() const'
/usr/bin/ld: CMakeFiles/qpid-proton-cpp.dir/src/connect_config.cpp.o: in function `proton::parse(Json::Value, proton:
:connection_options&)':
/root/irods/externals/qpid-proton0.36.0-0_src/qpid-proton/cpp/src/connect_config.cpp:168: undefined reference to `Jso
n::Value::asString() const'
/usr/bin/ld: CMakeFiles/qpid-proton-cpp.dir/src/connect_config.cpp.o: in function `proton::(anonymous namespace)::par
se_sasl(Json::Value, proton::connection_options&)':
/root/irods/externals/qpid-proton0.36.0-0_src/qpid-proton/cpp/src/connect_config.cpp:104: undefined reference to `Jso
n::Value::asString() const'
/usr/bin/ld: CMakeFiles/qpid-proton-cpp.dir/src/connect_config.cpp.o: in function `proton::(anonymous namespace)::get
_string(Json::Value const&, char const*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocato
r<char> > const&)':
/root/irods/externals/qpid-proton0.36.0-0_src/qpid-proton/cpp/src/connect_config.cpp:85: undefined reference to `Json::Value::Value(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)'
/usr/bin/ld: /root/irods/externals/qpid-proton0.36.0-0_src/qpid-proton/cpp/src/connect_config.cpp:85: undefined reference to `Json::Value::asString() const'
/usr/bin/ld: CMakeFiles/qpid-proton-cpp.dir/src/connect_config.cpp.o: in function `proton::apply_config(proton::connection_options&)':
/root/irods/externals/qpid-proton0.36.0-0_src/qpid-proton/cpp/src/connect_config.cpp:226: undefined reference to `Json::operator>>(std::__1::basic_istream<char, std::__1::char_traits<char> >&, Json::Value&)'
/usr/bin/ld: /root/irods/externals/qpid-proton0.36.0-0_src/qpid-proton/cpp/src/connect_config.cpp:230: undefined reference to `Json::operator>>(std::__1::basic_istream<char, std::__1::char_traits<char> >&, Json::Value&)'
/usr/bin/ld: CMakeFiles/qpid-proton-cpp.dir/src/connect_config.cpp.o: in function `proton::connect_config::parse(std::__1::basic_istream<char, std::__1::char_traits<char> >&, proton::connection_options&)':
/root/irods/externals/qpid-proton0.36.0-0_src/qpid-proton/cpp/src/connect_config.cpp:249: undefined reference to `Json::operator>>(std::__1::basic_istream<char, std::__1::char_traits<char> >&, Json::Value&)'
/usr/bin/ld: CMakeFiles/qpid-proton-cpp.dir/src/connect_config.cpp.o: in function `proton::connect_config::parse_default(proton::connection_options&)':
/root/irods/externals/qpid-proton0.36.0-0_src/qpid-proton/cpp/src/connect_config.cpp:278: undefined reference to `Json::operator>>(std::__1::basic_istream<char, std::__1::char_traits<char> >&, Json::Value&)'
clang-13: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [cpp/CMakeFiles/qpid-proton-cpp.dir/build.make:932: cpp/libqpid-proton-cpp.so.12.10.0] Error 1
make[2]: *** [CMakeFiles/Makefile2:2069: cpp/CMakeFiles/qpid-proton-cpp.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [Makefile:146: all] Error 2

```

**Comment by trel** [comment]
make sure you try `make qpid-proton_clean` first... in case you've been CMake cached from an earlier attempt.

**Comment by GeorgelT** [comment]
looked through the apt repo, there are only the externals in .deb form, is it possible to also get the irods-server and irods-commands packages. Should I make another issue for the main github repo?

I ran the test of trying to compile the main iRODS packages without that missing dependency, and it's past 70% currently. My first point still stands, would appreciate if .deb were available to skip this entire self compilation issue.

**Comment by trel** [comment]
We've never released iRODS server/core packages for deb11/el8/ub20 yet.   They're coming any day now with 4.3.0.

**Comment by trel** [comment]
The `make server` target in externals is what is needed for the server to build.

Other externals are for one/more separate plugins that require them.

**Comment by korydraughn** [comment]
Packages for the server and icommands will be available once work on 4.3.0 is complete. We are still tweaking/cleaning things up.

Like @trel said, 4.3.0 is coming. Until then, if you continue to experience trouble, try the following:
- Use a fresh docker container to build externals
- Use the [iRODS Development Environment](https://github.com/irods/irods_development_environment) to produce iRODS packages for the server and icommands
    - You can feed the development environment the externals packages you built

Using these methods have proven to be very stable and consistent. The core dev team uses these tools daily to produce various iRODS packages.
